### PR TITLE
Cube Sphere support

### PR DESCRIPTION
**Pull request by raphaeldussin**
@rabernat @brian-rose I wrote the low level function to read chunks for the Cube Sphere for issue #91. It works with limited capabilities and needs more testing.
Next steps:

- [x] add the CS configuration in the tests
- [x] add the CS geometry in the high level functions

### CODE REVIEW

**Comment by rabernat** [comment]
Any progress on this recently?

**Comment by brian-rose** [comment]
Not by me. I'm afraid I never found the time to understand enough about the xmitgcm internals to be able to offer any help here.

**Comment by raphaeldussin** [comment]
I got something working, just need to tidy it up

**Comment by AaronDavidSchneider** [comment]
Has there been any recent progress? I am not very familiar with xmitgcm and have just started to look into it, but I would love to use xarray!

**Comment by AaronDavidSchneider** [comment]
Hi,

just wanted to let you know that I can confirm that the following works for me:

```py
import xmitgcm

data_dir = "../run"
extra_metadata = xmitgcm.utils.get_extra_metadata(domain='cs', nx=6)
grid = xmitgcm.utils.get_grid_from_input(os.path.join(data_dir,"grid_cs32.face<NFACET>.bin"), geometry="cs", extra_metadata=extra_metadata)
```

Let me know if I can help you with anything or if there is a feature that I could help you with implementing.

**Comment by raphaeldussin** [comment]
@rabernat @AaronDavidSchneider I have made a few fixes to the code and I am now at the point where I can
load the grid metrics from the input binaries and load data using read_mds:

```python
import xmitgcm
import matplotlib.pylab as plt
import numpy as np
import xarray as xr
import os

home = os.environ['HOME']

# read grid from bin files
em = xmitgcm.utils.get_extra_metadata(domain='cs', nx=32)
grid = xmitgcm.utils.get_grid_from_input(f'{home}/.xmitgcm-test-data/grid_cs32/grid_cs32.face<NFACET>.bin', geometry='cs', extra_metadata=em)

grid['XC'].sel(face=1).plot()
plt.show()


# read file using read_mds

em = xmitgcm.utils.get_extra_metadata(domain='cs', nx=32)
raw = xmitgcm.utils.read_mds(f'{home}/.xmitgcm-test-data/global_oce_cs32/T.0000072020', chunks='CS', extra_metadata=em)

T = xr.DataArray(raw['T'], dims=('z','j','face','i'))

plt.figure()
#plt.pcolormesh(grid['XC'].sel(face=0), grid['YC'].sel(face=0), T.isel(z=0,face=0))
plt.pcolormesh(T.isel(z=0,face=0))
plt.colorbar()
plt.figure()
plt.pcolormesh(T.isel(z=0,face=1))
#plt.pcolormesh(grid['XC'].sel(face=1), grid['YC'].sel(face=1), T.isel(z=0,face=1))
plt.colorbar()
plt.figure()
plt.pcolormesh(T.isel(z=0,face=2))
#plt.pcolormesh(grid['XC'].sel(face=2), grid['YC'].sel(face=2), T.isel(z=0,face=2))
plt.colorbar()
plt.figure()
plt.pcolormesh(T.isel(z=0,face=3))
#plt.pcolormesh(grid['XC'].sel(face=3), grid['YC'].sel(face=3), T.isel(z=0,face=3))
plt.colorbar()
plt.figure()
plt.pcolormesh(T.isel(z=0,face=4))
#plt.pcolormesh(grid['XC'].sel(face=4), grid['YC'].sel(face=4), T.isel(z=0,face=4))
plt.colorbar()
plt.figure()
plt.pcolormesh(T.isel(z=0,face=5))
#plt.pcolormesh(grid['XC'].sel(face=5), grid['YC'].sel(face=5), T.isel(z=0,face=5))
plt.colorbar()

plt.show()
```

I am not sure about the orientation of the faces so this may need some revisiting and now someone
need to figure out the mds_store part.
I made a few changes in the testing, including adding some xfails in test_mds_store for features
that are not yet implemented.

Still a little bit of work that needs to be done, but I think I've left it in a better spot now.

**Comment by AaronDavidSchneider** [comment]
Thanks @raphaeldussin! I will have a deeper look into it on monday. 

At the moment I get an error for `read_mds` :
```
Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/pydevd.py", line 1448, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/Users/schneider/codes/MITgcm/verification/tutorial_held_suarez_cs/results/test_xmitgcm.py", line 19, in <module>
    raw = xmitgcm.utils.read_mds(f'{data_dir}/T.0000282240', chunks='CS', extra_metadata=em)
  File "/Users/schneider/codes/xmitgcm/xmitgcm/utils.py", line 215, in read_mds
    raise IOError("Cannot find the shape associated to %s in the \
OSError: Cannot find the shape associated to ../run/T.0000282240 in the                           metadata.
python-BaseException
``` 

I think I adapted the nescessary paths ...
I will look into it on monday.

**Comment by AaronDavidSchneider** [comment]
I can confirm that this works with the test dataset that you provide in your PR (should be removed from the tests folder?)
However, I believe its possible to have tiled output files (like the grid files: e.g. `T.0000276480.001.001.data`) which would brake `read_mds`.

**Comment by AaronDavidSchneider** [comment]
I think I found a simple approach to regrid the data. Attached you find the comparison of the cs plotting routine from mitgcmutils (which plots all tiles individually) and a regridded version (to longitude lattitude).
It all boils down to a oneliner basically...

I am not sure if this could help for `xmitgcm`. 

```py
import MITgcmutils as mit
from MITgcmutils.cs import pcol as cs_pcol
import matplotlib.pyplot as plt
import numpy as np

from scipy.interpolate import griddata

# import files:
# files should have shape 6n x n (e.g. 194x32, tiled data and coordinates)
x=mit.rdmds('../run/XG'); y=mit.rdmds('../run/YG'); e=mit.rdmds('../run/Eta',np.inf)

# plot using the pcol function of mitgcmutils
h = cs_pcol(x,y,e)
plt.savefig("test.png")
plt.show()

# get the corners of the current grid
x_max,x_min = np.max(x.flat), np.min(x.flat)
y_max,y_min = np.max(y.flat), np.min(y.flat)

# build the desired output grid
new_grid_x, new_grid_y = np.meshgrid(np.linspace(x_min,x_max,4*x.shape[1]),np.linspace(y_min,y_max,2*y.shape[1]))

# regrid data to 4n x 2n llt grid (e.g. 128x64)
# possible methods: linear, cubic, nearest (see documentation of scipy.interpolation.griddata)
new_e = griddata((x.flat,y.flat), e.flat, (new_grid_x,new_grid_y), method='cubic')

# plot using regrided data
plt.pcolor(new_grid_x,new_grid_y,new_e)
plt.savefig("test_regrid.png")
plt.show()

```

test
test_regrid

**Comment by raphaeldussin** [comment]
@AaronDavidSchneider remapping is the next step and for this I recommend using xESMF rather than griddata.

Any volunteers to finalize the PR? This means figuring out how to extend the high level mds_store to handle CS
and use read_mds with the correct arguments

**Comment by AaronDavidSchneider** [comment]
> @AaronDavidSchneider remapping is the next step and for this I recommend using xESMF rather than griddata.
> 
Thats a good idea in the context of xarrays. I didn't know about this library

> Any volunteers to finalize the PR? This means figuring out how to extend the high level mds_store to handle CS
> and use read_mds with the correct arguments
I will try to have a look into it. I got confused by the recursion last time I tried to understand the the `mds_store` function...

**Comment by raphaeldussin** [comment]
many thanks to @AaronDavidSchneider for figuring out the user-facing interface!!

**Comment by AaronDavidSchneider** [comment]
The current version seems to work only with 
```
useSingleCpuIO=.TRUE.
```
(first error is thrown in `_guess_model_horiz_dims`)

@rabernat: Is that a design choice?
I suppose this will be the same for the llc grid?

**Comment by raphaeldussin** [comment]
@AaronDavidSchneider I don't think xmitgcm can make sense of LLC or CS data written with `useSingleCpuIO=.FALSE.` since the IO layout is 2D and we lack the information to recombine into a 3D (face, y, x) array. Makes sense?

**Comment by AaronDavidSchneider** [comment]
> @AaronDavidSchneider I don't think xmitgcm can make sense of LLC or CS data written with `useSingleCpuIO=.FALSE.` since the IO layout is 2D and we lack the information to recombine into a 3D (face, y, x) array. Makes sense?

That makes sense. I suppose it would at least require additional assumptions.

@raphaeldussin, could you try to run the `tutorial_held_suarez_cs` verification? When I try to load the data using this PR, I get the following error (does not occur using the ocean dataset provided by you):

```
Traceback (most recent call last):
  File "/Users/schneider/codes/xmitgcm/xmitgcm/mds_store.py", line 636, in load_from_prefix
    extra_metadata=extra_metadata, chunks=chunks)
  File "/Users/schneider/codes/xmitgcm/xmitgcm/utils.py", line 290, in read_mds
    chunks=chunks)
  File "/Users/schneider/codes/xmitgcm/xmitgcm/utils.py", line 797, in read_all_variables
    use_mmap=use_mmap, use_dask=use_dask))
  File "/Users/schneider/codes/xmitgcm/xmitgcm/utils.py", line 848, in read_CS_chunks
    offset=0, order='C', partial_read=False)
  File "/Users/schneider/codes/xmitgcm/xmitgcm/utils.py", line 351, in read_raw_data
    actual_number_of_bytes))
OSError: File `/Users/schneider/codes/MITgcm/verification/tutorial_held_suarez_cs/run/XC.data` does not have the correct size (expected 1.76947e+06, found 49152)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/schneider/codes/xmitgcm/xmitgcm/mds_store.py", line 658, in load_from_prefix
    extra_metadata=extra_metadata, chunks=chunks)
  File "/Users/schneider/codes/xmitgcm/xmitgcm/utils.py", line 290, in read_mds
    chunks=chunks)
  File "/Users/schneider/codes/xmitgcm/xmitgcm/utils.py", line 797, in read_all_variables
    use_mmap=use_mmap, use_dask=use_dask))
  File "/Users/schneider/codes/xmitgcm/xmitgcm/utils.py", line 848, in read_CS_chunks
    offset=0, order='C', partial_read=False)
  File "/Users/schneider/codes/xmitgcm/xmitgcm/utils.py", line 351, in read_raw_data
    actual_number_of_bytes))
OSError: File `/Users/schneider/codes/MITgcm/verification/tutorial_held_suarez_cs/run/XC.data` does not have the correct size (expected 1.76947e+06, found 49152)
```

`tutorial_held_suarez_cs` uses a Binaryprecision of 64 while your dataset uses 32?! Is that the reason it fails? Shouldn't it work in both cases?

However, I am actually not really sure if this is even related to this PR. But if it is, I suppose we should fix this before merging this PR.

**Comment by raphaeldussin** [comment]
now includes fix for @AaronDavidSchneider for held suarez case

**Comment by AaronDavidSchneider** [comment]
Are we still missing anything on this PR @rabernat and @raphaeldussin?

**Comment by raphaeldussin** [comment]
@AaronDavidSchneider there's a few minor edits I want to make, mostly comments.

**Comment by AaronDavidSchneider** [comment]
Hi @raphaeldussin, I don't want to be a pain, but since this PR has come such a long way, it would be awesome to see it finalised. Let me know, if I can help you with anything.

**Comment by raphaeldussin** [comment]
it's all good for me. @rabernat are you happy with the PR?

**Comment by AaronDavidSchneider** [comment]
Hi @rabernat, just a quick friendly reminder (since it might have gotten lost).

File: xmitgcm/mds_store.py [hunk 1]
```diff
@@ -419,14 +430,19 @@ def __init__(self, data_dir, grid_dir=None,
...
 
         # --------------- LEGACY ----------------------
         if self.llc:
             if not user_metadata:
                 # if user didn't provide metadata, we default to llc
                 llc = get_extra_metadata(domain='llc', nx=self.nx)
                 extra_metadata = llc
+        if self.cs:
+            if not user_metadata:
+                # if user didn't provide metadata, we default to llc
```
Comment (rabernat) [inline xmitgcm/mds_store.py:443]: this comment seems incorrect

**Review by rabernat — APPROVED** [review]
Just reviewed; this looks great. Thanks @raphaeldussin!

**Pull request merged**